torch/autograd/_functions/linalg.py

```diff
@@ -1,71 +1,74 @@
 import torch
 
 from ..function import Function
+from ..variable import Variable
 
 
 class Diag(Function):
 
-    def __init__(self, diagonal_idx=0):
-        super(Diag, self).__init__()
-        self.diagonal_idx = diagonal_idx
+    @staticmethod
+    def forward(ctx, input, diagonal_idx=0):
+        ctx.diagonal_idx = diagonal_idx
+        return input.diag(ctx.diagonal_idx)
 
-    def forward(self, input):
-        return input.diag(self.diagonal_idx)
-
-    def backward(self, grad_output):
-        return grad_output.diag(self.diagonal_idx)
+    @staticmethod
+    def backward(ctx, grad_output):
+        return grad_output.diag(ctx.diagonal_idx), None
 
 
 class Tril(Function):
 
-    def __init__(self, diagonal_idx=0):
-        super(Tril, self).__init__()
-        self.diagonal_idx = diagonal_idx
-
-    def forward(self, input):
-        return input.tril(self.diagonal_idx)
+    @staticmethod
+    def forward(ctx, input, diagonal_idx=0):
+        ctx.diagonal_idx = diagonal_idx
+        return input.tril(ctx.diagonal_idx)
 
-    def backward(self, grad_output):
-        return grad_output.tril(self.diagonal_idx)
+    @staticmethod
+    def backward(ctx, grad_output):
+        return grad_output.tril(ctx.diagonal_idx), None
 
 
 class Triu(Function):
 
-    def __init__(self, diagonal_idx=0):
-        super(Triu, self).__init__()
-        self.diagonal_idx = diagonal_idx
-
-    def forward(self, input):
-        return input.triu(self.diagonal_idx)
+    @staticmethod
+    def forward(ctx, input, diagnoal_idx=0):
+        ctx.diagonal_idx = diagnoal_idx
+        return input.triu(ctx.diagonal_idx)
 
-    def backward(self, grad_output):
-        return grad_output.triu(self.diagonal_idx)
+    @staticmethod
+    def backward(ctx, grad_output):
+        return grad_output.triu(ctx.diagonal_idx), None
 
 
 class Trace(Function):
 
-    def forward(self, input):
-        self.isize = input.size()
-        return input.new((input.trace(),))
-
-    def backward(self, grad_output):
-        isize = self.isize
-        grad_input = grad_output.new(isize).zero_()
-        grad_input.view(-1)[::(isize[1] + 1)] = grad_output[0]
+    @staticmethod
+    def forward(ctx, input):
+        ctx.isize = input.size()
+        return input.new((input.trace(), ))
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        isize = ctx.isize
+        mask = grad_output.data.new(isize).zero_()
+        mask.view(-1)[::(isize[1] + 1)] = 1
+        mask = Variable(mask)
+        grad_output_expand = grad_output.view(1, 1).expand(isize)
+        grad_input = mask * grad_output_expand
```

Comment:
It will raise error `RuntimeError: inconsistent tensor size`
In fact, I tried to implement it using the old way at first.
```python
class Trace(Function):

    @staticmethod
    def forward(ctx, input):
        ctx.isize = input.size()
        return input.new((input.trace(), ))

    @staticmethod
    def backward(ctx, grad_output):
        isize = ctx.isize
        min_size = min(isize)
        grad_input = Variable(grad_output.data.new(isize).zero_())
        grad_input.view(-1)[::(isize[1] + 1)] = grad_output.expand(min_size)
        return grad_input
```
It raise another error
```
RuntimeError: in-place operations can be only used on variables that don't share storage with any other variables, but detected that there are 2 objects sharing it
```